### PR TITLE
Fix erroneous massive acceleration in the wrong direction in grsim motion controller

### LINKED ISSUE #270: Robot moves away from desired destination grsim
**Describe the bug**
When simulating movement in grsim, if the robot is set to track and follow the ball, in certain situations the robot will travel in the exact opposite direction of the ball. This occurs when the ball is placed in the opposite direction of the robots current velocity. The robot will move onto the correct vector to be lined up with the ball, however instead of moving towards it, it moves in the opposite direction.

**To Reproduce**
1. In the grsim communication code set up the robot to follow the ball.
2. Run the ai and launch grsim.
3. Place the ball anywhere on the field to get the robot moving.
4. Replace the ball behind the robot.
5. See bug.

**Expected behavior**
The robot should turn around and head towards the ball.

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

This fixes the case in the grsim motion controller where the robot would massively accelerate in the wrong direction when the robot receives a small acceleration in the opposite direction that exceeds the additional velocity in the correct direction applied during that tick. 

### Testing Done

Robot no longer accelerates in the opposite direction when hitting the wall in grsim. Unit test added for this particular case. 5 test cases still fail - working on it right now. 

### Resolved Issues

Resolves #270 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [ ] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by garethellis0** [comment]
Just to double check, is this still WIP? Or did you want a review?

**Comment by jordanzliu** [comment]
Yeah, this is done and ready for review

File: src/thunderbots/software/grsim_communication/main.cpp [hunk 1]
```diff
@@ -91,6 +93,8 @@ int main(int argc, char** argv)
         // Clear all primitives each tick
         primitives.clear();
 
+        primitives.emplace_back(
+            new MovePrimitive(0, ball.position(), Angle::full(), 0.5));
```
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/main.cpp]: I don't think we should need the changes to this file?

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp [hunk 2]
```diff
@@ -14,9 +14,10 @@
 #include "grsim_communication/motion_controller/motion_controller.h"
 
 #include <algorithm>
+#include <chrono>
```
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: Can we remove this?

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp [hunk 3]
```diff
@@ -84,29 +85,21 @@ Vector MotionController::determineLinearVelocity(const Robot robot, const Point
...
+    Vector unit_vector_to_dest = (dest - robot.position()).norm();
 
-    // Calculate the vector required to counteract our movement perpendicular to the
-    // vector towards the dest. This "pushes" us onto a velocity vector straight towards
-    // the destination
+    // Calculate the vector of our velocity perpendicular to the vector toward the
+    // destination
     Vector robot_velocity_perpendicular_to_dest =
-        robot.velocity() - robot_velocity_towards_dest;
+        robot.velocity().dot(unit_vector_to_dest.perp()) * unit_vector_to_dest.perp();
```
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp:93]: What's up with the last `* unit_vector_to_dest.perp();` here? I understand the first bit, but after that you should have the perpendicular component of the motion, no?

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp [hunk 4]
```diff
@@ -120,10 +113,24 @@ Vector MotionController::determineLinearVelocity(const Robot robot, const Point
...
-    new_robot_velocity_magnitude = std::clamp<double>(new_robot_velocity_magnitude, 0,
+
+    // https://github.com/UBC-Thunderbots/Software/issues/270
+    // Check for the case where we are moving away from the target AND the additional
+    // velocity is greater in magnitude than the current velocity, as the above
+    // function will increase in magnitude as the distance to the destination increases
+    if ((robot.velocity().orientation() - unit_vector_to_dest.orientation())
+                .angleMod()
+                .abs() > Angle::quarter() &&
+        additional_velocity.len() < robot.velocity().len())
```
Comment (garethellis0) [inline src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: Could we break this out into a few boolean variables? I think that would make this much more readable.

File: src/thunderbots/software/test/grsim_backend/motion_controller.cpp [hunk 5]
```diff
@@ -694,6 +694,29 @@ TEST_F(MotionControllerTest, negative_rotation_position_test)
...
+    Point destination       = Point(0.1, 0);
+    Angle destination_angle = Angle::ofRadians(0);
+
+    MotionController::Velocity robot_velocities =
+        MotionController::bangBangVelocityController(robot, destination, 2,
+                                                     destination_angle, delta_time);
+    // if the destination is on the +x side of the robot, the velocity should always have
+    // x > 0
+    EXPECT_GT(robot_velocities.linear_velocity.x(), initial_velocity.x());
+}
```
Comment (garethellis0) [inline src/thunderbots/software/test/grsim_backend/motion_controller.cpp:719]: Just as a sanity check, could we add a test for the case where we have a large velocity away from the destination?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp:93]: the dot product is the length of the projection of `robot.velocity()` onto `unit_vector_to_dest.perp()`, and multiplying by `unit_vector_to_dest.perp()` gives the vector that is the image of `robot.velocity()` onto `unit_vector_to_dest.perp()`, which is what we are looking for.

File: src/thunderbots/software/grsim_communication/main.cpp (hunk 1 above)
Comment (jordanzliu) [reply src/thunderbots/software/grsim_communication/main.cpp]: testing code, removed

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 2 above)
Comment (jordanzliu) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: that was part of my jank csv output, removed

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 4 above)
Comment (jordanzliu) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: broke out the first condition into a bool, 2nd condition is pretty self-explanatory

File: src/thunderbots/software/test/grsim_backend/motion_controller.cpp (hunk 5 above)
Comment (jordanzliu) [reply src/thunderbots/software/test/grsim_backend/motion_controller.cpp:719]: added

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp:93]: ah of course. whoops!

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: Awesome! Could we rename `moving_in_opposite_direction` to `moving_in_opposite_direction_from_target` just to be super clear?

File: src/thunderbots/software/test/grsim_backend/motion_controller.cpp [hunk 6]
```diff
@@ -694,6 +694,47 @@ TEST_F(MotionControllerTest, negative_rotation_position_test)
                 POSITION_TOLERANCE);
 }
 
+TEST_F(MotionControllerTest, wrong_direction_impulse_while_close_to_destination_test)
```
Comment (garethellis0) [inline src/thunderbots/software/test/grsim_backend/motion_controller.cpp]:
``` suggestion
TEST_F(MotionControllerTest, wrong_direction_impulse_while_close_to_destination_small_initial_velocity_test)
```

File: src/thunderbots/software/test/grsim_backend/motion_controller.cpp [hunk 7]
```diff
@@ -694,6 +694,47 @@ TEST_F(MotionControllerTest, negative_rotation_position_test)
...
+
+    MotionController::Velocity robot_velocities =
+        MotionController::bangBangVelocityController(robot, destination, 2,
+                                                     destination_angle, delta_time);
+    // if the destination is on the +x side of the robot, the velocity should always have
+    // x > 0
+    EXPECT_GT(robot_velocities.linear_velocity.x(), initial_velocity.x());
+}
+
+TEST_F(MotionControllerTest, wrong_direction_impulse_while_close_to_destination_2_test)
```
Comment (garethellis0) [inline src/thunderbots/software/test/grsim_backend/motion_controller.cpp]:
``` suggestion
TEST_F(MotionControllerTest, wrong_direction_impulse_while_close_to_destination_large_initial_velocity_test)
```

File: src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp (hunk 4 above)
Comment (jordanzliu) [reply src/thunderbots/software/grsim_communication/motion_controller/motion_controller.cpp]: renamed to `moving_away_from_dest`

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**